File: agentscope-service/frontend/src/api/teams.ts
/*
 * Copyright 2024-2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { api } from '@/lib/apiClient';

export type TeamPhase = 'Pending' | 'Running' | 'Idle' | 'Completed' | 'Failed' | string;
export type MemberPhase = 'Joining' | 'Working' | 'Idle' | 'Lost' | 'Failed' | 'Shutdown' | string;
export type TaskState = 'pending' | 'in_progress' | 'completed' | 'failed' | string;

export interface Team {
  id?: string;
  name: string;
  namespace: string;
  objective: string;
  phase: TeamPhase;
  leadRef?: string;
  leadPrompt?: string;
  startedAt?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface TeamMember {
  memberName: string;
  agentRef: string;
  prompt?: string;
  planApproval?: boolean;
  planText?: string;
  planStatus?: 'pending' | 'approved' | 'rejected' | string;
  origin?: string;
  deployMode?: 'managed' | 'byo' | string;
  managedAgentId?: string;
  ownerId?: string;
  phase: MemberPhase;
  sessionId?: string;
  managedSessionId?: string;
  instanceRef?: string;
  currentTask?: string;
  restartCount?: number;
  lastRestartReason?: string;
}

export interface TeamTask {
  taskId: string;
  teamName?: string;
  namespace?: string;
  subject: string;
  description?: string;
  state: TaskState;
  owner?: string;
  blockedBy?: string[];
  result?: string;
  version: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface TeamMessage {
  id: string;
  fromMember: string;
  toMember: string;
  content: string;
  kind?: string;
  delivered?: boolean;
  createdAt?: string;
}

export interface TeamTaskSummary {
  total: number;
  pending: number;
  inProgress: number;
  completed: number;
}

export interface TeamCreateLead {
  agentRef: string;
  prompt?: string;
  deployMode?: 'managed' | 'byo';
  managedAgentId?: string;
  ownerId?: string;
}

export interface TeamCreateMember {
  name: string;
  agentRef: string;
  prompt?: string;
  planApproval?: boolean;
  deployMode?: 'managed' | 'byo';
  managedAgentId?: string;
  ownerId?: string;
}

export interface TeamCreateRequest {
  name: string;
  namespace?: string;
  objective: string;
  lead: TeamCreateLead;
  members?: TeamCreateMember[];
}

function nsQuery(namespace?: string) {
  const ns = namespace || 'default';
  return `namespace=${encodeURIComponent(ns)}`;
}

export function listTeams(namespace?: string) {
  const q = namespace ? `?${nsQuery(namespace)}` : '';
  return api.get<{ items: Team[] }>(`/api/v1/teams${q}`);
}

export function getTeam(teamName: string, namespace = 'default') {
  return api.get<{
    team: Team;
    members: TeamMember[];
    tasks: TeamTaskSummary;
  }>(`/api/v1/teams/${encodeURIComponent(teamName)}?${nsQuery(namespace)}`);
}

export function createTeam(body: TeamCreateRequest) {
  return api.post<{ team: Team; members: TeamMember[] }>('/api/v1/teams', body);
}

export function deleteTeam(teamName: string, namespace = 'default') {
  return api.delete(`/api/v1/teams/${encodeURIComponent(teamName)}?${nsQuery(namespace)}`);
}

export function completeTeam(teamName: string, namespace = 'default') {
  return api.post<{ team: Team }>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/complete?${nsQuery(namespace)}`,
    {},
  );
}

export function removeTeamMember(teamName: string, memberName: string, namespace = 'default') {
  return api.delete(
    `/api/v1/teams/${encodeURIComponent(teamName)}/members/${encodeURIComponent(memberName)}?${nsQuery(namespace)}`,
  );
}

export function unclaimTeamTask(
  teamName: string,
  taskId: string,
  resourceVersion?: number,
  namespace = 'default',
) {
  return api.post<TeamTask>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/unclaim?${nsQuery(namespace)}`,
    {
      resourceVersion: resourceVersion != null ? String(resourceVersion) : undefined,
    },
  );
}

export function submitMemberPlan(
  teamName: string,
  memberName: string,
  planText: string,
  namespace = 'default',
) {
  return api.post(
    `/api/v1/teams/${encodeURIComponent(teamName)}/members/${encodeURIComponent(memberName)}/plan?${nsQuery(namespace)}`,
    { planText },
  );
}

export function approveMemberPlan(teamName: string, memberName: string, namespace = 'default') {
  return api.post(
    `/api/v1/teams/${encodeURIComponent(teamName)}/members/${encodeURIComponent(memberName)}/plan/approve?${nsQuery(namespace)}`,
    {},
  );
}

export function rejectMemberPlan(teamName: string, memberName: string, namespace = 'default') {
  return api.post(
    `/api/v1/teams/${encodeURIComponent(teamName)}/members/${encodeURIComponent(memberName)}/plan/reject?${nsQuery(namespace)}`,
    {},
  );
}

export function listTeamMembers(teamName: string, namespace = 'default') {
  return api.get<{ lead?: TeamMember; members: TeamMember[] }>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/members?${nsQuery(namespace)}`,
  );
}

export function spawnTeamMember(
  teamName: string,
  body: TeamCreateMember,
  namespace = 'default',
) {
  return api.post<{ member: TeamMember; status: string }>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/members?${nsQuery(namespace)}`,
    body,
  );
}

export function listTeamTasks(teamName: string, namespace = 'default') {
  return api.get<{ tasks: TeamTask[]; summary: TeamTaskSummary }>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/tasks?${nsQuery(namespace)}`,
  );
}

export function createTeamTask(
  teamName: string,
  body: { subject: string; description?: string; blockedBy?: string[]; owner?: string },
  namespace = 'default',
) {
  return api.post<TeamTask>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/tasks?${nsQuery(namespace)}`,
    body,
  );
}

export function assignTeamTask(
  teamName: string,
  taskId: string,
  owner: string,
  resourceVersion?: number,
  namespace = 'default',
) {
  return api.post<TeamTask>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/assign?${nsQuery(namespace)}`,
    {
      owner,
      resourceVersion: resourceVersion != null ? String(resourceVersion) : undefined,
    },
  );
}

export function claimTeamTask(
  teamName: string,
  taskId: string,
  claimedBy: string,
  resourceVersion?: number,
  namespace = 'default',
) {
  return api.post<TeamTask>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/claim?${nsQuery(namespace)}`,
    {
      claimedBy,
      resourceVersion: resourceVersion != null ? String(resourceVersion) : undefined,
    },
  );
}

export function completeTeamTask(
  teamName: string,
  taskId: string,
  result?: string,
  namespace = 'default',
) {
  return api.post<TeamTask>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/tasks/${encodeURIComponent(taskId)}/complete?${nsQuery(namespace)}`,
    { result },
  );
}

export function listTeamMessages(teamName: string, limit = 50, namespace = 'default') {
  return api.get<{ messages: TeamMessage[] }>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/messages?${nsQuery(namespace)}&limit=${limit}`,
  );
}

export function sendTeamMessage(
  teamName: string,
  body: { from: string; to: string; content: string },
  namespace = 'default',
) {
  return api.post<TeamMessage>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/messages?${nsQuery(namespace)}`,
    body,
  );
}

export function listTeamEvents(teamName: string, after = '', limit = 50, namespace = 'default') {
  const params = new URLSearchParams({
    namespace,
    limit: String(limit),
  });
  if (after) params.set('after', after);
  return api.get<{ events: TeamMessage[]; after?: string }>(
    `/api/v1/teams/${encodeURIComponent(teamName)}/events?${params}`,
  );
}

export function teamPhaseTone(
  phase?: string,
): 'default' | 'success' | 'warning' | 'danger' | 'info' {
  switch ((phase || '').toLowerCase()) {
    case 'running':
    case 'working':
      return 'warning';
    case 'completed':
    case 'idle':
      return 'success';
    case 'pending':
    case 'joining':
      return 'info';
    case 'failed':
    case 'lost':
      return 'danger';
    default:
      return 'default';
  }
}

export function chatSessionPath(member: TeamMember): string | null {
  const id = managedChatSessionId(member);
  if (!id) return null;
  return `/sessions/${encodeURIComponent(id)}`;
}

/** Managed member session id for in-page chat (null for BYO / unbound). */
export function managedChatSessionId(member: TeamMember): string | null {
  if (member.deployMode !== 'managed') return null;
  const id = member.managedSessionId || member.sessionId;
  if (!id) return null;
  return id;
}

File: agentscope-service/aistio/internal/store/memory/teams.go
// Copyright 2024-2026 the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package memory

import (
	"context"
	"time"

	"github.com/spring-ai-alibaba/aistio/internal/store"
)

type teamRepo struct{ s *Store }

func (r *teamRepo) Create(_ context.Context, team *store.Team) (*store.Team, error) {
	r.s.mu.Lock()
	defer r.s.mu.Unlock()
	key := teamKey(team.Namespace, team.Name)
	if _, ok := r.s.teams[key]; ok {
		return nil, store.ErrConflict
	}
	now := time.Now().UTC()
	cp := *team
	cp.ID = nextID(&r.s.nextTeamID)
	if cp.Phase == "" {
		cp.Phase = store.TeamPhasePending
	}
	cp.CreatedAt = now
	cp.UpdatedAt = now
	r.s.teams[key] = &cp
	out := cp
	return &out, nil
}

func (r *teamRepo) Get(_ context.Context, namespace, name string) (*store.Team, error) {
	r.s.mu.RLock()
	defer r.s.mu.RUnlock()
	t, ok := r.s.teams[teamKey(namespace, name)]
	if !ok {
		return nil, store.ErrNotFound
	}
	cp := *t
	return &cp, nil
}

func (r *teamRepo) List(_ context.Context, namespace string) ([]*store.Team, error) {
	r.s.mu.RLock()
	defer r.s.mu.RUnlock()
	var out []*store.Team
	for _, t := range r.s.teams {
		if namespace != "" && t.Namespace != namespace {
			continue
		}
		cp := *t
		out = append(out, &cp)
	}
	return out, nil
}

func (r *teamRepo) UpdatePhase(_ context.Context, namespace, name, phase string) error {
	r.s.mu.Lock()
	defer r.s.mu.Unlock()
	t, ok := r.s.teams[teamKey(namespace, name)]
	if !ok {
		return store.ErrNotFound
	}
	t.Phase = phase
	t.UpdatedAt = time.Now().UTC()
	if phase == store.TeamPhaseRunning && t.StartedAt == nil {
		now := time.Now().UTC()
		t.StartedAt = &now
	}
	return nil
}

func (r *teamRepo) Update(_ context.Context, team *store.Team) (*store.Team, error) {
	r.s.mu.Lock()
	defer r.s.mu.Unlock()
	key := teamKey(team.Namespace, team.Name)
	cur, ok := r.s.teams[key]
	if !ok {
		return nil, store.ErrNotFound
	}
	cp := *team
	cp.ID = cur.ID
	cp.CreatedAt = cur.CreatedAt
	cp.UpdatedAt = time.Now().UTC()
	r.s.teams[key] = &cp
	out := cp
	return &out, nil
}

func (r *teamRepo) Delete(_ context.Context, namespace, name string) error {
	r.s.mu.Lock()
	defer r.s.mu.Unlock()
	key := teamKey(namespace, name)
	if _, ok := r.s.teams[key]; !ok {
		return store.ErrNotFound
	}
	delete(r.s.teams, key)
	kept := r.s.teamMembers[:0]
	for _, m := range r.s.teamMembers {
		if m.Namespace == namespace && m.TeamName == name {
			continue
		}
		kept = append(kept, m)
	}
	r.s.teamMembers = kept
	return nil
}

func (r *teamRepo) UpsertMember(_ context.Context, m *store.TeamMember) (*store.TeamMember, error) {
	r.s.mu.Lock()
	defer r.s.mu.Unlock()
	now := time.Now().UTC()
	for i := range r.s.teamMembers {
		cur := &r.s.teamMembers[i]
		if cur.Namespace == m.Namespace && cur.TeamName == m.TeamName && cur.MemberName == m.MemberName {
			id, created := cur.ID, cur.CreatedAt
			*cur = *m
			cur.ID = id
			cur.CreatedAt = created
			cur.UpdatedAt = now
			if cur.Phase == "" {
				cur.Phase = store.MemberPhaseJoining
			}
			if cur.Origin == "" {
				cur.Origin = store.MemberOriginStatic
			}
			if cur.DeployMode == "" {
				cur.DeployMode = store.MemberDeployBYO
			}
			cp := *cur
			return &cp, nil
		}
	}
	cp := *m
	cp.ID = nextID(&r.s.nextMemberID)
	if cp.Phase == "" {
		cp.Phase = store.MemberPhaseJoining
	}
	if cp.Origin == "" {
		cp.Origin = store.MemberOriginStatic
	}
	if cp.DeployMode == "" {
		cp.DeployMode = store.MemberDeployBYO
	}
	cp.CreatedAt = now
	cp.UpdatedAt = now
	r.s.teamMembers = append(r.s.teamMembers, cp)
	out := cp
	return &out, nil
}

func (r *teamRepo) GetMember(_ context.Context, namespace, teamName, memberName string) (*store.TeamMember, error) {
	r.s.mu.RLock()
	defer r.s.mu.RUnlock()
	for i := range r.s.teamMembers {
		m := r.s.teamMembers[i]
		if m.Namespace == namespace && m.TeamName == teamName && m.MemberName == memberName {
			cp := m
			return &cp, nil
		}
	}
	return nil, store.ErrNotFound
}

func (r *teamRepo) ListMembers(_ context.Context, namespace, teamName string) ([]*store.TeamMember, error) {
	r.s.mu.RLock()
	defer r.s.mu.RUnlock()
	var out []*store.TeamMember
	for i := range r.s.teamMembers {
		m := r.s.teamMembers[i]
		if m.Namespace == namespace && m.TeamName == teamName {
			cp := m
			out = append(out, &cp)
		}
	}
	return out, nil
}

func (r *teamRepo) RemoveMember(_ context.Context, namespace, teamName, memberName string) error {
	r.s.mu.Lock()
	defer r.s.mu.Unlock()
	kept := r.s.teamMembers[:0]
	found := false
	for _, m := range r.s.teamMembers {
		if m.Namespace == namespace && m.TeamName == teamName && m.MemberName == memberName {
			found = true
			continue
		}
		kept = append(kept, m)
	}
	r.s.teamMembers = kept
	if !found {
		return store.ErrNotFound
	}
	return nil
}

func (r *teamRepo) BindMemberSession(_ context.Context, namespace, teamName, memberName, sessionID, managedSessionID, instanceRef string) error {
	r.s.mu.Lock()
	defer r.s.mu.Unlock()
	for i := range r.s.teamMembers {
		m := &r.s.teamMembers[i]
		if m.Namespace == namespace && m.TeamName == teamName && m.MemberName == memberName {
			m.SessionID = sessionID
			m.ManagedSessionID = managedSessionID
			m.InstanceRef = instanceRef
			m.UpdatedAt = time.Now().UTC()
			return nil
		}
	}
	return store.ErrNotFound
}

func (r *teamRepo) UpdateMemberPhase(_ context.Context, namespace, teamName, memberName, phase string) error {
	r.s.mu.Lock()
	defer r.s.mu.Unlock()
	for i := range r.s.teamMembers {
		m := &r.s.teamMembers[i]
		if m.Namespace == namespace && m.TeamName == teamName && m.MemberName == memberName {
			m.Phase = phase
			m.UpdatedAt = time.Now().UTC()
			return nil
		}
	}
	return store.ErrNotFound
}

func (r *teamRepo) FindMemberBySessionID(_ context.Context, sessionID string) (*store.TeamMember, error) {
	if sessionID == "" {
		return nil, store.ErrNotFound
	}
	r.s.mu.RLock()
	defer r.s.mu.RUnlock()
	var best *store.TeamMember
	for i := range r.s.teamMembers {
		m := &r.s.teamMembers[i]
		if m.SessionID == sessionID || m.ManagedSessionID == sessionID {
			cp := *m
			if best == nil || cp.UpdatedAt.After(best.UpdatedAt) {
				best = &cp
			}
		}
	}
	if best == nil {
		return nil, store.ErrNotFound
	}
	return best, nil
}

func teamKey(namespace, name string) string {
	return namespace + "/" + name
}
